src/com/google/ProfilingReport.java

```diff
@@ -0,0 +1,249 @@
+
+    List<Map<String, List<ProfilingResult>>> listOfProfilingResultMaps = new ArrayList<>();
+    for (FileProfilingResult fileProfilingResult : result) {
+      listOfProfilingResultMaps.add(fileProfilingResult.profilingDataPerFunction());
+    }
+
+    this.percentOfFunctionsExecuted = calculatePercentOfInstrumentationExecuted(
+        listOfProfilingResultMaps, (t) -> t.equals(InstrumentationType.FUNCTION));
+    this.percentOfBranchesExecuted = calculatePercentOfInstrumentationExecuted(
+        listOfProfilingResultMaps, (t) -> t.equals(InstrumentationType.BRANCH) || t
+            .equals(InstrumentationType.BRANCH_DEFAULT));
+
+  }
+
+  /**
+   * A class which groups profiling results by source file and contains a mapping of them to each
+   * function which is present in that source file. Used also for Json conversion.
+   */
+  @AutoValue
+  abstract static class FileProfilingResult {
+
+    static FileProfilingResult create(String fileName, float percentOfFunctionsExecuted,
+        float percentOfBranchesExecuted,
+        Map<String, List<ProfilingResult>> profilingDataPerFunction) {
+      return new AutoValue_ProfilingReport_FileProfilingResult(fileName, percentOfFunctionsExecuted,
+          percentOfBranchesExecuted, profilingDataPerFunction);
+    }
+
+    abstract String fileName();
+
+    abstract float percentOfFunctionsExecuted();
+
+    abstract float percentOfBranchesExecuted();
+
+    abstract Map<String, List<ProfilingResult>> profilingDataPerFunction();
+  }
+
+  /**
+   * This class contains a detailed breakdown of each instrumentation point. For any encoded param,
+   * this class will contain its details. The `executed` property will be a percent of total times
+   * the instrumentation point was called across all provided instrumentation reports. i.e if
+   * this.executed = 100, then every report would have this point present. Similarly, ProfilingData
+   * takes the average of all ProfilingData's for this instrumentation point in all reports. Used
+   * also for Json conversion.
+   */
+  @AutoValue
+  abstract static class ProfilingResult {
+
+    static ProfilingResult create(InstrumentationMapping instrumentationMapping, String param,
+        float executed, ProfilingData data) {
+      return new AutoValue_ProfilingReport_ProfilingResult(param,
+          instrumentationMapping.getType(param), instrumentationMapping.getLineNo(param),
+          instrumentationMapping.getColNo(param), executed, data);
+    }
+
+    abstract String param();
+
+    abstract InstrumentationType type();
+
+    abstract int lineNo();
+
+    abstract int colNo();
+
+    abstract float executed();
+
+    abstract ProfilingData data();
+
+  }
+
+
+  /**
+   * The instrumentation will send the report in a JSON format where the JSON is a dictionary of the
+   * encoded params to an object of the data collected. Initially this will just be the frequency.
+   */
+  public static class ProfilingData {
```

Comment:
Can we also make this AutoValue?

It might make the class definition slightly larger for now, but will be good if we need to update the class in the future.

Reply:
I'm not sure if I am not using AutoValue incorrectly, but marking this class as AutoValue will break JSON deserialization. This class is used to convert the executionResults.json file for reference.

Reply:
Interesting, that's fine then.

Could you just add a short comment explaining why this can't be AutoValue?